src/nanoFramework.IoT.Device.CodeConverter/GenericsTemplates/Span.cs

```diff
@@ -0,0 +1,186 @@
+namespace System
+{
+    /// <summary>
+    /// Provides a type- and memory-safe representation of a contiguous region of arbitrary array.
+    /// </summary>
+    [Serializable, CLSCompliant(false)]
+    public readonly ref struct Span<T>
+    {
+        private readonly T[] _array;
+        private readonly int _start;
+        private readonly int _length;
+
+        /// <summary>
+        /// Creates a new Span object over the entirety of a specified array.
+        /// </summary>
+        /// <param name="array">The array from which to create the System.Span object.</param>
+        public Span(T[] array)
+        {
+            _array = array;
+            _length = array?.Length ?? 0;
+            _start = 0;
+        }
+
+        /// <summary>
+        /// Creates a new Span object that includes a specified number of elements
+        /// of an array starting at a specified index.
+        /// </summary>
+        /// <param name="array">The source array.</param>
+        /// <param name="start">The index of the first element to include in the new System.Span</param>
+        /// <param name="length">The number of elements to include in the new System.Span</param>
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// array is null, but start or length is non-zero. -or- start is outside the bounds
+        /// of the array. -or- start and length exceeds the number of elements in the array.
+        /// </exception>
+        public Span(T[] array, int start, int length)
+        {
+            if (array != null)
+            {
+                if ((start + length > array.Length) || (start >= array.Length))
+                {
+                    // Array length too small
+                    throw new ArgumentOutOfRangeException();
+                }
+            }
+            else
+            {
+                if ((start != 0) || (length != 0))
+                {
+                    // Array is null but start and length are not 0
+                    throw new ArgumentOutOfRangeException();
+                }
+            }
+
+            _array = array;
+            _start = start;
+            _length = length;
+        }
+
+        /// <summary>
+        /// Gets the element at the specified zero-based index.
+        /// </summary>
+        /// <param name="index">The zero-based index of the element.</param>
+        /// <returns>The element at the specified index.</returns>
+        // public ref T this[int index] => ref _array[_start + index]; // <= this is not working and raises exception after few access
+        public T this[int index]
+        {
+            get
+            {
+                int realIndex = _start + index;
```

Comment:
the get should be:
```
if (index &gt; _length)                
{                    
throw new ArgumentOutOfRangeException();                }
}
return _array[_start + index];
```

Reply:
Ah, I see. Length is not length of the array.

Reply:
By the way it's `>=`, not `>`

Reply:
Good catch for this one. Mind a PR on msborlib? :-)